jetstream.go

```diff
@@ -0,0 +1,334 @@
+	if j.ctx == nil && j.timeout == 0 {
+		j.timeout = dftl
+	}
+
+	if j.ctx != nil {
+		return context.WithCancel(j.ctx)
+	}
+
+	return context.WithTimeout(context.Background(), j.timeout)
+}
+
+// AckOption configures the various JetStream message acknowledgement helpers
+type AckOption func(opts *jsOpts) error
+
+// PublishOption configures publishing messages
+type PublishOption func(opts *jsOpts) error
+
+// PublishExpectsStream waits for an ack after publishing and ensure it's from a specific stream, empty arguments waits for any valid acknowledgement
+func PublishExpectsStream(stream ...string) PublishOption {
+	return func(opts *jsOpts) error {
+		switch len(stream) {
+		case 0:
+			opts.ack.str = ""
+
+		case 1:
+			opts.ack.str = stream[0]
+			if !isValidJSName(opts.ack.str) {
+				return ErrInvalidStreamName
+			}
+
+		default:
+			return ErrInvalidStreamName
+		}
+
+		return nil
+	}
+}
+
+// PublishStreamTimeout sets the period of time to wait for JetStream to acknowledge receipt, defaults to JetStreamTimeout option
+func PublishStreamTimeout(t time.Duration) PublishOption {
+	return func(opts *jsOpts) error {
+		opts.timeout = t
+		return nil
+	}
+}
+
+// PublishCtx sets an interrupt context for waiting on a stream to reply
+func PublishCtx(ctx context.Context) PublishOption {
+	return func(opts *jsOpts) error {
+		opts.ctx = ctx
+		return nil
+	}
+}
+
+// AckWaitDuration waits for confirmation from the JetStream server
+func AckWaitDuration(d time.Duration) AckOption {
+	return func(opts *jsOpts) error {
+		opts.timeout = d
+		return nil
+	}
+}
+
+func (m *Msg) jsAck(body []byte, opts ...AckOption) error {
+	if m.Reply == "" {
+		return ErrMsgNoReply
+	}
+
+	if m == nil || m.Sub == nil {
+		return ErrMsgNotBound
+	}
+
+	m.Sub.mu.Lock()
+	nc := m.Sub.conn
+	m.Sub.mu.Unlock()
+
+	var err error
+
+	aopts := newJsOpts()
+	for _, f := range opts {
+		err = f(aopts)
+		if err != nil {
+			return err
+		}
+	}
+
+	if aopts.timeout == 0 {
+		return m.Respond(body)
+	}
+
+	_, err = nc.Request(m.Reply, body, aopts.timeout)
+
+	return err
+}
+
+// Ack acknowledges a JetStream messages received from a Consumer, indicating the message
+// should not be received again later
+func (m *Msg) Ack(opts ...AckOption) error {
+	return m.jsAck(AckAck, opts...)
+}
+
+// Nak acknowledges a JetStream message received from a Consumer, indicating that the message
+// is not completely processed and should be sent again later
+func (m *Msg) Nak(opts ...AckOption) error {
+	return m.jsAck(AckNak, opts...)
+}
+
+// AckProgress acknowledges a Jetstream message received from a Consumer, indicating that work is
+// ongoing and further processing time is required equal to the configured AckWait of the Consumer
+func (m *Msg) AckProgress(opts ...AckOption) error {
+	return m.jsAck(AckProgress, opts...)
+}
+
+// AckNext performs an Ack() and request that the next message be sent to subject ib
+func (m *Msg) AckNext(ib string) error {
```

Comment:
I think we need option here to for batch size as well. Also need to figure out overload for ack the ack and the next message if present.

Reply:
Yeah I thought of that, but we dont actually process batch size on AckNext

```go
case bytes.Equal(msg, AckNext):
		o.ackMsg(sseq, dseq, dcount)
		o.processNextMsgReq(nil, nil, subject, reply, nil)
```

while the signature for `processNextMsgReq` is `func (o *Consumer) processNextMsgReq(_ *subscription, _ *client, _, reply string, msg []byte) {` so we never send the message on, no pulling out of payload size.  This may well be a bug in JetStream but that's why I didnt add it here since its not supportd.

Reply:
Ah might be a not implemented feature..

Reply:
What is `ib string` for here?

Reply:
where to send the data, `subject` be better name

Reply:
But you return a msg from the call itself no?

Reply:
from `AckNext`

Reply:
not following at all, lets zoom later

Reply:
Ignore that, too early in am for me. Apologies.

Reply:
Should ackNext return the next message or an err if not available to match `Request`?

Reply:
I dont think so, I believe typical case would be to feed a buffer or existing subscription.

Reply:
I would have thought main use case was workqueue and you just pull next work item, process, ack and get next one and repeat.

Reply:
```go
nc.Subscribe(ib, func(m *nats.Msg) {
   // do stuff
   m.AckNext(ib)
})
```

more typically you might have a worker pull consuming messages off a queue, and the AckNext() would request the next message into the queue.

I'd rather add an additional function here if you want it to be more like a request for the next message

Reply:
I think you are right, what would that look like? For the `Request()` like `AckNext()`

Reply:
`msg, err := m.AckFetchNext(opts...)` or `s/Fetch/Request/`

Reply:
`AckAndFetch()`?

Reply:
k, will implement and we can see how it looks

Reply:
Added `AckAndFetch()` https://github.com/nats-io/nats.go/pull/585/files#diff-5ad60403adccb2da7f1010ef809aaa9aR278

Reply:
ok what do you think?

Reply:
seems fine